### PR TITLE
Breakout sessions

### PR DESCRIPTION
**Pull request by Coread**
# COMPLETES SPARK-344982

## This pull request addresses

This implements breakout sessions for attendees.

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested

< ENUMERATE TESTS PERFORMED, WHETHER MANUAL OR AUTOMATED >

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [ ] I discussed changes with code owners prior to submitting this pull request

- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [x] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: webpack.config.js [hunk 1]
```diff
@@ -62,11 +62,11 @@ module.exports = (
         // we don't need to manually rebuild after changing code.
         alias[`./packages/${packageName}`] = path.resolve(
           __dirname,
-          `./packages/${packageName}/src/index.js`
+          `./packages/${packageName}/src/index`
```
Comment (Coread) [inline webpack.config.js:65]: needed so a typescript plugin can be used in the sample apps

File: packages/@webex/plugin-meetings/src/breakouts/README.md [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+// This will cause the participant to return to the main session
+breakouts.currentBreakoutSession.leave()
+
+// Ask for help when in a breakout session. 10 second cooldown.
+breakout.currentBreakoutSession.askForHelp()
+```
+
+#### Members
+
+The current session members are available on the meeting object as normal
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/breakouts/README.md:56]: Maybe rephrase this, hard to understand what you mean.

Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/breakouts/README.md:56]: I mean the whole paragrah/s? Can't tell

File: packages/@webex/plugin-meetings/src/breakouts/breakout.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    if (!mainSession) {
+      throw new Error('Cannot leave, no main session found');
+    }
+
+    return mainSession.join();
+  },
+
+  /**
+   * Sends a help request for the current breakout
+   */
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/breakouts/breakout.ts:88]: @returns {Promise}

File: packages/@webex/plugin-meetings/src/breakouts/breakout.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+      body: {
+        groupId: this.groupId,
+        sessionId: this.sessionId
+      }
+    });
+  },
+
+  /**
+   * Parses the participants from the locus object
+   * @param locus
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/breakouts/breakout.ts:102]: What is locus, object, or url?

File: packages/@webex/plugin-meetings/src/breakouts/collection.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+ * Copyright (c) 2015-2023 Cisco Systems, Inc. See LICENSE file.
+ */
+
+import AmpCollection from 'ampersand-collection';
+
+import {MEETINGS} from '../constants';
+
+import Breakout from './breakout';
+
+const BreakoutCollection = AmpCollection.extend({
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/breakouts/collection.ts:11]: What is this `AmpCollection` dependency, and why does it help us here?

File: packages/@webex/plugin-meetings/src/breakouts/index.ts [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+  },
+
+  /**
+   * The initial roster lists need to be quried because you don't
+   * get a breakout.roster event when you join the meeting
+   */
+  queryRosters() {
+    this.webex.request(
+      {uri: `${this.url}/roster`, qs: {locusUrl: btoa(this.locusUrl)}
+    }).then((result) => {
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/breakouts/index.ts:95]: should we handle a .catch ?

File: packages/@webex/plugin-meetings/src/breakouts/index.ts [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+    }
+
+    session.parseRoster(locus);
+  },
+
+  /**
+   * Sets up listener for broadcast messages sent to the breakout session
+   */
+  listenToBroadcastMessages() {
+    this.listenTo(this.webex.internal.llm, 'event:breakout.message', (event) => {
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/breakouts/index.ts:122]: why not just `this.webex.internal.llm.on('event:breakout.message', ...)`

File: packages/@webex/plugin-meetings/src/breakouts/index.ts [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+      const {data: {senderUserId, sentTime, message}} = event;
+
+      this.trigger(BREAKOUTS.EVENTS.MESSAGE, {
+        senderUserId,
+        sentTime,
+        message,
+        // FIXME: This is only the current sessionId
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/breakouts/index.ts:129]: is this because the new llm logic you added initially automatically connects to the new dataChannelURL and this could happen half way through this action (receiving the `event:breakout.message` and triggering it further)?

File: packages/@webex/plugin-meetings/src/breakouts/index.ts [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+        sessionId: this.currentBreakoutSession.sessionId,
+      });
+    })
+  },
+
+  /**
+   * Sets up a listener for roster messags from mecury
+   */
+  listenToBreakoutRosters() {
+    this.listenTo(this.webex.internal.mercury, 'event:breakout.roster', (event) => {
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/breakouts/index.ts:141]: Any reason we get the roster events on mercury and not LLM?

File: packages/@webex/plugin-meetings/src/breakouts/index.ts [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+    })
+  },
+
+  /**
+   * Updates the information about the current breakout
+   * @param {Object} params
+   * @returns {void}
+   */
+  updateBreakout(params) {
+    this.set(params);
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/breakouts/index.ts:153]: why is this needed?

File: packages/@webex/plugin-meetings/src/breakouts/index.ts [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+        forEach(payload.breakoutSessions[state], (breakout) => {
+          const {sessionId} = breakout;
+
+          if (!breakouts[sessionId]) {
+            breakouts[sessionId] = breakout;
+            breakouts[sessionId].active = false;
+            breakouts[sessionId].assigned = false;
+            breakouts[sessionId].allowed = false;
+            breakouts[sessionId].assignedCurrent = false;
+            breakouts[sessionId].requested = false;
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/breakouts/index.ts:189]: why are you setting this to false for all states?
doesn't the `breakout` object already have information about its own state?

File: packages/@webex/plugin-meetings/src/breakouts/index.ts [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+          }
+
+          breakouts[sessionId][state] = true;
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/breakouts/index.ts:192]: oh, I see what you're doing. 
You're updating it here...

Seems a bit confusing?

you could just do something like 
```js
 breakouts[sessionId].allowed = state === 'allowed';
```
 in the if statement above.
 
 Also, would be nice to create constants for all the available states.

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 13]
```diff
@@ -1188,6 +1199,49 @@ export default class Meeting extends StatelessWebexPlugin {
...
+    });
+
+    this.breakouts.on(BREAKOUTS.EVENTS.MEMBERS_UPDATE, () => {
+      Trigger.trigger(
+        this,
+        {
+          file: 'meeting/index',
+          function: 'setUpBreakoutsListener'
+        },
+        EVENT_TRIGGERS.MEETING_BREAKOUTS_UPDATE,
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:1242]: one says members update(in the breakounts.on), and the other general breakouts update(in the trigger). Is this on purpose?

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 14]
```diff
@@ -1712,6 +1766,19 @@ export default class Meeting extends StatelessWebexPlugin {
         }
       });
 
+    this.locusInfo.on(LOCUSINFO.EVENTS.CONTROLS_MEETING_BREAKOUT_UPDATED,
+      ({breakout}) => {
+        this.breakouts.updateBreakout(breakout);
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:1771]: do we need to call this on `this.breakouts.on(BREAKOUTS.EVENTS.MEMBERS_UPDATE, () => {` too?

File: packages/@webex/plugin-meetings/src/meeting/request.ts [hunk 15]
```diff
@@ -112,6 +115,10 @@ export default class MeetingRequest extends StatelessWebexPlugin {
       }
     };
 
+    if (breakoutsSupported) {
+      body.deviceCapabilities = ['BREAKOUTS_SUPPORTED'];
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/meeting/request.ts:119]: make it a constant?

**Review by torgeadelin — CHANGES_REQUESTED** [review]
I've tested manually in the samples app and seems to work ok.
Most of my comments are questions, with small request of changes.

File: packages/@webex/plugin-meetings/src/member/util.ts [hunk 16]
```diff
@@ -120,6 +120,18 @@ MemberUtil.isHandRaised = (participant: any) => {
   return participant.controls?.hand?.raised || false;
 };
 
+/**
+ * @param {Object} participant the locus participant
+ * @returns {Boolean}
+ */
+MemberUtil.isSupportsBreakouts = (participant) => {
```
[the remark this answers is not in the document]
Comment (Coread) [reply packages/@webex/plugin-meetings/src/member/util.ts]: changed to isBreakoutsSupported

File: packages/@webex/plugin-meetings/package.json [hunk 17]
```diff
@@ -53,6 +53,7 @@
     "@webex/plugin-people": "workspace:^",
```
[the remark this answers is not in the document]
Comment (Coread) [reply packages/@webex/plugin-meetings/package.json:53]: I reset these to the version from beta but they are still showing up here. Perhaps github is not showing the diff correctly?

File: packages/@webex/plugin-meetings/src/breakouts/README.md (hunk 2 above)
Comment (Coread) [reply packages/@webex/plugin-meetings/src/breakouts/README.md:56]: I've updated this paragraph. Hopefully it is clearer now

File: packages/@webex/plugin-meetings/src/breakouts/breakout.ts (hunk 4 above)
Comment (Coread) [reply packages/@webex/plugin-meetings/src/breakouts/breakout.ts:102]: locus object. I'll update the comment

File: packages/@webex/plugin-meetings/src/breakouts/collection.ts (hunk 5 above)
Comment (Coread) [reply packages/@webex/plugin-meetings/src/breakouts/collection.ts:11]: We use ampersand state in the SDK. Ampersand state has the concept of collections, groups of models. The collections offer helpful methods for searching, sorting models as well as event propagation. From what I can tell looking at the rest of the SDK, in places where authors didn't use ampersand collections, they had to resort to writing a lot of extra custom cost to do things like a get a model by its ID or to listen to an event within a model in a collection. There's no need for all that extra code, since the framework we already chose does this for us.

File: packages/@webex/plugin-meetings/src/breakouts/index.ts [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+   * Update the current locus url of the meeting
+   */
+  locusUrlUpdate(locusUrl) {
+    this.set('locusUrl', locusUrl);
+  },
+
+  /**
+   * The initial roster lists need to be quried because you don't
+   * get a breakout.roster event when you join the meeting
+   */
```
[the remark this answers is not in the document]
Comment (Coread) [reply packages/@webex/plugin-meetings/src/breakouts/index.ts:91]: fixed

File: packages/@webex/plugin-meetings/src/breakouts/breakout.ts (hunk 3 above)
Comment (Coread) [reply packages/@webex/plugin-meetings/src/breakouts/breakout.ts:88]: updated

File: packages/@webex/plugin-meetings/src/breakouts/breakout.ts [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+        groupId: this.groupId,
+        sessionId: this.sessionId
+      }
+    });
+  },
+
+  /**
+   * Leaves the breakout session to return to the main session
+   * @returns {Promise}
+   */
```
[the remark this answers is not in the document]
Comment (Coread) [reply packages/@webex/plugin-meetings/src/breakouts/breakout.ts:71]: updated

File: packages/@webex/plugin-meetings/src/breakouts/index.ts (hunk 7 above)
Comment (Coread) [reply packages/@webex/plugin-meetings/src/breakouts/index.ts:122]: This is again where taking advantage of the framework that ampersand state provides means you can worry less and write less code. `listenTo` is designed so that when you destroy the object you automatically stop listening to events. You don't have to worry that you've left the event handler there and need to clean it up manually.

File: packages/@webex/plugin-meetings/src/breakouts/index.ts (hunk 8 above)
Comment (Coread) [reply packages/@webex/plugin-meetings/src/breakouts/index.ts:129]: Yes. I am happy to leave this edge case at this stage though. It would only be a problem in cases where the application developer was filtering the broadcast messages in their app by sessionId and dropped a message because it wasn't for the current room.

File: packages/@webex/plugin-meetings/src/breakouts/index.ts (hunk 9 above)
Comment (Coread) [reply packages/@webex/plugin-meetings/src/breakouts/index.ts:141]: I don't know the specific reasoning architecturally. If I had to guess it would be because the main meeting roster list arrives via the locus DTO on normal mercury so this is to be consistent with that.

File: packages/@webex/plugin-meetings/src/breakouts/index.ts (hunk 10 above)
Comment (Coread) [reply packages/@webex/plugin-meetings/src/breakouts/index.ts:153]: You mean why not call .set directly? In the rest of the sdk we typically don't access built in methods of ampersand state directly. After all, ampersand state is basically an implementation choice and so exposing its API to something outside of the unit might make it more brittle. Typically we define methods like updateWhatever and then other parts of the code don't have to know that set exists, or how it works. Adding a separate method also gives us room to sanitize input in the future

File: packages/@webex/plugin-meetings/src/breakouts/index.ts (hunk 11 above)
Comment (Coread) [reply packages/@webex/plugin-meetings/src/breakouts/index.ts:189]: I'm initializing as false and then updating the one to be true later

File: packages/@webex/plugin-meetings/src/breakouts/index.ts (hunk 12 above)
Comment (Coread) [reply packages/@webex/plugin-meetings/src/breakouts/index.ts:192]: The payload arrives as an object with arrays of sessions in different states
e.g.
{
active: [{session1}, {session2}],
allowed: [{session1}]
}

I want to flatten and deduplicate this into a single collection of models where each state is just a flag. Sessions can appear in multiple places, so you can't just assign say allowed based on whether the current bucket is allowed because you may be overwriting a value set in by a previous bucket.

Comment (Coread) [reply packages/@webex/plugin-meetings/src/breakouts/index.ts:192]: I'm now using constants for session states

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 13 above)
Comment (Coread) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:1242]: Are you asking why this is the same event being triggered eventually? This is because the result is the same for the application developer. Something has changed in the breakouts and they need to go update their state. There was no point in propagating two distinct events outside of the meetings plugin for this.

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 14 above)
Comment (Coread) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:1771]: no. This is receiving information from locus and updating the breakouts with that information. After doing that it triggers the event so the app consuming this knows something has changed. The other event is listening to changes that have already happened regarding the members of the breakouts. No need to update anything else on the breakouts, simply to inform the consuming app that something has changed.

File: packages/@webex/plugin-meetings/src/meeting/request.ts (hunk 15 above)
Comment (Coread) [reply packages/@webex/plugin-meetings/src/meeting/request.ts]: updated

File: packages/@webex/plugin-meetings/src/breakouts/collection.ts (hunk 5 above)
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/breakouts/collection.ts:11]: Thanks for the thorough explanation! Makes sense!

File: packages/@webex/plugin-meetings/src/breakouts/index.ts (hunk 6 above)
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/breakouts/index.ts:95]: I guess not?

File: packages/@webex/plugin-meetings/src/breakouts/index.ts (hunk 7 above)
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/breakouts/index.ts:122]: thank you for the explanation.

File: packages/@webex/plugin-meetings/src/breakouts/index.ts (hunk 8 above)
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/breakouts/index.ts:129]: yeah, don't mind it here, was just trying to understand why was needed.

File: packages/@webex/plugin-meetings/src/breakouts/index.ts (hunk 10 above)
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/breakouts/index.ts:153]: I was referring to why calling this.set() since it's called below again with more parameters. I guess it would be nice to describe what is `params` or how does the object look like.

File: packages/@webex/plugin-meetings/src/breakouts/index.ts (hunk 12 above)
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/breakouts/index.ts:192]: makes sense!!

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 13 above)
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:1243]: I see, I was just curious why we are listening on "member_update" and trigger "breakout_update" instead of "member_update". Is a bit confusing but since it's the same data then maybe it's fine? Would be nice to be documented.

**Review by torgeadelin — APPROVED** [review]
LGTM

File: packages/@webex/plugin-meetings/src/breakouts/collection.ts (hunk 5 above)
Comment (arun3528) [reply packages/@webex/plugin-meetings/src/breakouts/collection.ts:11]: we will need the same for members at some point

File: packages/@webex/plugin-meetings/src/breakouts/index.ts (hunk 6 above)
Comment (Coread) [reply packages/@webex/plugin-meetings/src/breakouts/index.ts:95]: I actually missed this one. I'll take a look. We typically do log an error in this situation.

File: packages/@webex/plugin-meetings/src/breakouts/index.ts (hunk 10 above)
Comment (Coread) [reply packages/@webex/plugin-meetings/src/breakouts/index.ts:153]: yes, perhaps we could define a type somehow to show what exactly gets set on the object. Ampersand state ignores the params that are not defined anyway.

File: packages/@webex/plugin-meetings/src/breakouts/index.ts (hunk 6 above)
Comment (Coread) [reply packages/@webex/plugin-meetings/src/breakouts/index.ts]: I've added a catch here

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 13 above)
Comment (Coread) [reply packages/@webex/plugin-meetings/src/meeting/index.ts]: The current session sends member update events and those will still apply. This is for members of other sessions, sessions the user isn't in which is handled by the consuming app going over the members of those sessions. It felt overkill to submit individual member update events for each other breakout session and then force the consuming app to keep track of all of those listeners, given that they all come from a single roster update event.

**Review by arun3528 — APPROVED** [review]
